hail/src/main/scala/is/hail/expr/ir/ForwardLets.scala

```diff
@@ -0,0 +1,45 @@
+package is.hail.expr.ir
+
+object ForwardLets {
+  def extract(x: IR): (String, IR, IR) = (x: @unchecked) match {
+    case Let(name, value, body) => (name, value, body)
+    case AggLet(name, value, body) => (name, value, body)
+  }
+
+  private def rewriteIR(ir0: IR): IR = {
+    val UsesAndDefs(uses, _) = ComputeUsesAndDefs(ir0)
+    val nestingDepth = NestingDepth(ir0)
+
+    def rewrite(ir: IR, env: Env[IR]): IR = {
+      ir match {
+        case _: Let | _: AggLet =>
+          val (name, value, body) = extract(ir)
+          val refs = uses.lookup(ir)
+
+          if (value.isInstanceOf[Ref] ||
+            refs.isEmpty ||
+            (refs.size == 1 && nestingDepth.lookup(refs.head) == nestingDepth.lookup(ir)))
+            rewrite(body, env.bind(name -> value))
+          else
+            MapIR(rewrite(_, env))(ir)
+        case x@Ref(name, _) =>
+          env.lookupOption(name) match {
+            case Some(fwd) => fwd
+            case None => x
+          }
+        case _ =>
+          MapIR(rewrite(_, env))(ir)
+      }
+    }
+
+    rewrite(ir0, Env.empty)
+  }
+
+  def apply(ir0: BaseIR, needsCopy: Boolean = true): BaseIR =
```

Comment:
You mean copy every time, or copy never?

Reply:
I guess I would say two things:
 - copy every time since that's basically what you're doing,
 - we should never never share subtrees so there's should be no need to copy.
Of course, we do share subtrees so just copy for now, but we need to fix that ASAP if we're going to be using this ref equality model for analyses.